test/e2e/framework/metrics/metrics_grabber.go

```diff
@@ -102,37 +123,48 @@ func NewMetricsGrabber(c clientset.Interface, ec clientset.Interface, kubelets b
 			break
 		}
 	}
-	if kubeScheduler == "" {
-		scheduler = false
-		klog.Warningf("Can't find kube-scheduler pod. Grabbing metrics from kube-scheduler is disabled.")
-	}
-	if kubeControllerManager == "" {
-		controllers = false
-		klog.Warningf("Can't find kube-controller-manager pod. Grabbing metrics from kube-controller-manager is disabled.")
-	}
-	if snapshotControllerManager == "" {
-		snapshotController = false
-		klog.Warningf("Can't find snapshot-controller pod. Grabbing metrics from snapshot-controller is disabled.")
-	}
-	if ec == nil {
+	if clusterAutoscaler && ec == nil {
 		klog.Warningf("Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.")
 	}
 
 	return &Grabber{
 		client:                     c,
 		externalClient:             ec,
+		config:                     config,
 		grabFromAPIServer:          apiServer,
-		grabFromControllerManager:  controllers,
+		grabFromControllerManager:  checkPodDebugHandlers(c, controllers, "kube-controller-manager", kubeControllerManager),
 		grabFromKubelets:           kubelets,
-		grabFromScheduler:          scheduler,
+		grabFromScheduler:          checkPodDebugHandlers(c, scheduler, "kube-scheduler", kubeScheduler),
 		grabFromClusterAutoscaler:  clusterAutoscaler,
-		grabFromSnapshotController: snapshotController,
+		grabFromSnapshotController: checkPodDebugHandlers(c, snapshotController, "snapshot-controller", snapshotControllerManager),
 		kubeScheduler:              kubeScheduler,
 		kubeControllerManager:      kubeControllerManager,
 		snapshotController:         snapshotControllerManager,
 	}, nil
 }
 
+func checkPodDebugHandlers(c clientset.Interface, requested bool, component, podName string) bool {
+	if !requested {
+		return false
+	}
+	if podName == "" {
+		klog.Warningf("Can't find %s pod. Grabbing metrics from %s is disabled.", component, component)
+		return false
+	}
+
+	// The debug handlers on the host where the pod runs might be disabled.
+	// We can check that indirectly by trying to retrieve log output.
+	limit := int64(1)
+	if _, err := c.CoreV1().Pods(metav1.NamespaceSystem).GetLogs(podName, &v1.PodLogOptions{LimitBytes: &limit}).DoRaw(context.TODO()); err != nil {
```

Comment:
is there any possible race here that the pod is listed, handlers are available but we fail to get the logs?
is there an specific error we can match against?

Reply:
This could fail because the containers aren't running yet (IMHO unlikely, because this is now getting called after the e2e.test binary has already ensured that the cluster is up) or simply because of load.

There is some error text about debug handlers that we could check for and we could retry for other errors. I decided against it because its more complex, unlikely to happen (admittedly a bit subjective) and the impact of incorrectly disabling metrics grabbing is low. Run a job often enough and most of the time it should execute the tests.